backend/compilers/download.py

```diff
@@ -154,97 +99,96 @@ def get_compiler(
-            )
-            return
-        # image_digest missing or out of date, so pull
-        logger.debug("%s has newer image available; pulling ...", full_docker_image)
-        client_manager.pull(full_docker_image)
-
-    try:
-        container = client_manager.create_container(full_docker_image)
-    except Exception as err:
-        logger.error("Unable to create container for %s: %s", full_docker_image, err)
-        return
-
-    try:
-        compiler_container_path = f"/compilers/{platform_id}/{compiler_id}"
-        stream, _ = container.get_archive(compiler_container_path)
-    except Exception as err:
-        logger.error(
-            "Unable to pull '%s' from container: %s", compiler_container_path, err
+    if image_digest.exists() and image_digest.read_text() == remote_image_digest and not force:
+        logger.debug(
+            "%s image is present and at latest version, skipping!", compiler_id
         )
-        container.remove()
         return
 
-    try:
-        with tempfile.TemporaryDirectory() as temp_dir:
-            tar_file = Path(temp_dir) / f"{platform_id}_{compiler_id}.tar"
-            logger.debug("Writing to %s...", tar_file)
-            with tar_file.open("wb") as f:
-                for chunk in stream:
-                    f.write(chunk)
+    # First, get a token to do our operations with
+    token = get_token(docker_registry, github_repo, docker_image)
 
-            logger.debug("Extracting %s to %s", tar_file, download_cache)
-            shutil.unpack_archive(tar_file, download_cache)
-
-            if compiler_dir.exists():
-                logger.debug("Removing existing compiler dir: %s", compiler_dir)
-                shutil.rmtree(compiler_dir)
+    # Then, get the container image index. This will give us all the
+    # container images associated with this tag. There may be different
+    # images for different OSes/platforms. In our case, we only expect one
+    # though.
+    image_url = (
+        f"https://{docker_registry}/v2/{github_repo}/{docker_image}/manifests/{tag}"
+    )
+    headers = {
+        "Accept": "application/vnd.oci.image.index.v1+json",
+        "Authorization": f"Bearer {token}",
+    }
+    resp = requests.get(image_url, headers=headers, timeout=10)
+    if resp.status_code != 200:
+        return None
+    data = resp.json()
+
+    # Grab the first image manifest, and download it.
+    manifest = next((manifest for manifest in data['manifests'] if manifest['mediaType'] == 'application/vnd.oci.image.manifest.v1+json'))
+    digest = manifest['digest']
+    mime = manifest['mediaType']
+    headers['Accept'] = mime
+    url = f"https://{docker_registry}/v2/{github_repo}/{docker_image}/manifests/{digest}"
+    resp = requests.get(url, headers=headers, timeout=10)
+    if resp.status_code != 200:
+        return None
+    data = resp.json()
+
+    # Then, find the layer, and download it.
+    layer = data['layers'][0]
```

Comment:
does this assume that the compiler image is only going to have a single layer (currently true, but not sure if it will always be the case)?

Reply:
Yes. I could adjust this to extract the layers one after the other though.

Reply:
Implemented this.